docs(otel): add contrib collector to self-managed Elastic guide (#7032)
elastic/docs-content: solutions/observability/get-started/opentelemetry/use-cases/upstream-collector/index.md

Reviewer: I don’t think this guide warrants an `index.md` in a separate folder within the OpenTelemetry use cases, but I may be wrong. I see the pattern, but I don’t necessarily agree with it.

Author: I used that for consistency with kubernetes/ and llms/, as this guide may grow. In fact, I have another PR cooking that relies on this folder structure: https://github.com/elastic/docs-content/pull/7061

---
navigation_title: Upstream OpenTelemetry Collector
description: Send data from an upstream OpenTelemetry Collector to a self-managed Elastic Stack by routing it through an EDOT Collector gateway.
applies_to:

>   stack: ga 9.2+

Reviewer: If this guide is specifically for sending data to a self-managed cluster, then I think we should use the [deployment dimension](https://www.elastic.co/docs/contribute-docs/how-to/cumulative-docs/guidelines#dimensions) for the `applies_to`:

```suggestion
applies_to:
  deployment:
    self: ga 9.2+
```

Author: Right, I keep forgetting that. Thanks!

products:
  - id: observability
  - id: edot-collector
---

# Send data from an upstream OpenTelemetry Collector [upstream-collector-self-managed]

> This guide shows how to forward telemetry data from an existing (upstream) OpenTelemetry Collector to a self-managed {{stack}} using an EDOT Collector configured as a gateway. The examples use the [contrib distribution](https://github.com/open-telemetry/opentelemetry-collector-releases/tree/main/distributions/otelcol-contrib) (`otelcol-contrib`), but the same approach works with any OTel Collector distribution, including vendor distributions and custom builds assembled with the [OpenTelemetry Collector Builder](https://opentelemetry.io/docs/collector/custom-collector/).

Reviewer: Can we expand the EDOT abbreviation and link to the relevant docs?

Author: Ah yes, another thing I keep forgetting is expanding the abbreviation 😅 Done & done

## When to use this setup

> Use this pattern if you:

Reviewer: I don’t think _pattern_ is the right word here 🤔

Author: changed

* Already run an existing OpenTelemetry Collector and want to add Elastic as a backend without replacing your current setup

> * Need to fan out telemetry to multiple observability backends from a single Collector

Reviewer: Sounds like a jargon expression, maybe:
```suggestion
* Need to send telemetry to multiple observability backends from a single Collector
```

* Evaluate Elastic alongside another backend before committing to a full migration

> * Use a technology or language for which Elastic doesn't provide an EDOT distribution

Reviewer: > EDOT distribution

Tautology?

Author: Ehhh, I don't think it's that much of a tautology, and we do use "EDOT distribution" in some of our docs, but we can be more specific and just call it "EDOT SDK"

Specific = better anyway

## Architecture

> Your services send telemetry to the contrib Collector (`otelcol-contrib`), which forwards it over OTLP/gRPC to the EDOT Collector gateway. The gateway applies Elastic-specific processing and writes directly to {{es}}.

Reviewer: we could make this slightly more general:

```suggestion
Your services send telemetry to an OpenTelemetry Collector (for example, `otelcol-contrib`), which forwards it over OTLP/gRPC to the EDOT Collector gateway. The gateway applies Elastic-specific processing and writes directly to {{es}}.
```

```mermaid
flowchart LR
    S["Your services"] -->|OTLP| U["Upstream otelcol-contrib<br/>(one per host)"]
    U -->|"OTLP/gRPC :4317"| G

    subgraph G["EDOT Collector (gateway)"]
        P["elasticapm processor<br/>enriches spans"]
        C["elasticapm connector<br/>generates aggregated metrics"]
    end

>     G -->|"elasticsearch exporter<br/>mapping.mode: otel"| E[("ES")]

Reviewer: Maybe this would be better:

```suggestion
    G -->|"elasticsearch exporter<br/>mapping.mode: otel"| E[("Elasticsearch")]
```

```

> The `elasticsearch` exporter with `mapping.mode: otel` is the recommended path for self-managed deployments. The Managed OTLP endpoint isn't available for self-managed installations. Sending directly to {{apm-server-or-mis}} through OTLP is also possible but not recommended. Use the EDOT gateway path when available.

Reviewer: This paragraph is a bit awkward here without additional context (also it sounds more like it should be a note?).

Not sure users know they had any other options in the first place. Also, we’re introducing the managed OTLP endpoint in the text without any further information about it… We need a link to the mOTLP docs at least.

Author: Changed to a note and added the link. Good point BTW, it does read as more of a note now that I think about it

> ## Prerequisites

Reviewer: We could use the [how-to guide](https://www.elastic.co/docs/contribute-docs/content-types/how-tos) content type as the base structure for consistency across the docs.

> * A running self-managed {{es}} cluster

Reviewer: This could use a link to https://www.elastic.co/docs/deploy-manage/deploy/self-managed

> * The [EDOT Collector](elastic-agent://reference/edot-collector/index.md) installed on the gateway host. It ships as part of the {{agent}} package and runs as {{agent}} in `otel` mode.

Reviewer: The EDOT abbreviation needs to be expanded at first use though

* An existing OpenTelemetry Collector installed on your agent hosts. This guide uses [`otelcol-contrib`](https://opentelemetry.io/docs/collector/installation/).
* Network connectivity from your Collector hosts to the EDOT gateway host on port 4317

> ::::{stepper}

Reviewer: We need a heading before the stepper. Currently the steps appear as headings under Prerequisites.

Author: true

:::{step} Create an {{es}} API key

The EDOT gateway authenticates to {{es}} using an API key.

> 1. In {{kib}}, navigate to **{{stack-manage-app}}** → **API keys**.

Reviewer: I don't think this wording aligns with our guidance on [referring to apps and pages](https://www.elastic.co/docs/contribute-docs/style-guide/ui-writing#referring-to-apps-and-pages).

2. Select **Create API key**.

> 3. Give the key a name (for example, `edot-gateway`) and assign it the necessary privileges for writing to {{es}} data streams.

Reviewer: > necessary privileges

What are those? We should be more specific. Alternatively, if the process for getting an API key is described in another doc, you could pull this into the _Prerequisites_/_Before you begin_ section as a requirement, but still explicitly mentioning what privileges must be given for the API key

Author: Yeah, I updated it to be more specific, good point Visha

> 4. Copy the encoded key to use as the `ELASTIC_API_KEY` value in the gateway configuration.

Reviewer: maybe:

```suggestion
4. Copy the encoded key to use as the value of the `ELASTIC_API_KEY` environment variable in the gateway configuration.
```

:::

:::{step} Configure the EDOT gateway

Create a configuration file for the EDOT Collector acting as a gateway. Save this as `gateway.yml` on the gateway host.

Set the following environment variables on the gateway host before starting the Collector:

```bash
export ELASTIC_ENDPOINT=https://your-elasticsearch:9200
export ELASTIC_API_KEY=your-encoded-api-key
```

> Then create `gateway.yml`:

Reviewer: OK, I’m confused, but maybe it’s the wording here... On line 65, we say:

> Create a configuration file for the EDOT Collector acting as a gateway. Save this as `gateway.yml` on the gateway host.

And now we create it?

Author: remnants of it being 2 different steps I think... I rephrased to avoid the awkward duplication

      processors: [elasticapm]
      exporters: [elasticapm, elasticsearch/otel]
    metrics:
      receivers: [otlp]
      processors: []
      exporters: [elasticsearch/otel]
    metrics/aggregated-otel-metrics:
      receivers: [elasticapm]
      processors: []
      exporters: [elasticsearch/otel]
    logs:
      receivers: [otlp]
      processors: []
      exporters: [elasticsearch/otel]
```

Key components in this configuration:

* **`elasticapm` processor** (under `processors`): Enriches spans with attributes required by the {{product.apm}} UI.
* **`elasticapm` connector** (under `connectors`): Generates pre-aggregated {{product.apm}} metrics from trace data. It appears as an exporter in the `traces` pipeline and as a receiver in the `metrics/aggregated-otel-metrics` pipeline.
* **`elasticsearch/otel` exporter**: Writes data directly to {{es}} using native OpenTelemetry data streams (`mapping.mode: otel`). The exporter handles batching automatically using `sending_queue`. Refer to [Performance and batching](elastic-agent://reference/edot-collector/components/elasticsearchexporter.md#performance-and-batching) to customize throughput for your environment.

:::{note}

> The `elasticapm` connector and processor are required for full {{product.apm}} functionality (service maps, transaction histograms, service-level indicators). You only need them when exporting directly to {{es}}. If you send to the Managed OTLP endpoint or {{apm-server-or-mis}}, they are not required.

Reviewer: ```suggestion
The `elasticapm` connector and processor are required for full {{product.apm}} functionality (service maps, transaction histograms, service-level indicators). You only need them when exporting directly to {{es}}. If you send data to the Managed OTLP endpoint or {{apm-server-or-mis}}, they are not required.
```

Refer to [{{product.apm}} services missing due to misconfigured elasticapm connector](/troubleshoot/ingest/opentelemetry/edot-collector/misconfigured-elasticapm-connector.md) for more information.
:::

> Start the EDOT gateway. The EDOT Collector is the {{agent}} binary run in `otel` mode, so start it with the `otel` subcommand:

Reviewer: Something with the stepper formatting is causing this paragraph to land outside of the step. I think you need to increase the `:` in the `:::{step}` directive so that it doesn’t mess with the closing `:::` of the `:::{note}` directive. The current closing `:::` of the note probably closes the step as well.

Also, maybe we could have numbered sub-steps in this step section?

1. Create a configuration file…
2. Set the following environment variables…
3. Start the EDOT gateway…

```bash
elastic-agent otel --config gateway.yml
```

:::

:::{step} Configure the contrib Collector

On each contrib Collector host, configure the OTLP exporter to point to the EDOT gateway. Add or update the `exporters` and `service` sections in your existing `config.yml`:

```yaml
exporters:
  otlp:
    endpoint: "gateway-host:4317"
    tls:

>       insecure: true  # Set to `false` and configure `ca_file` for production

Reviewer: `ca_file` is mentioned for the first time here. Where should that be configured? We should either explain, or link to some specific docs. Maybe we can add a separate comment line in the snippet that shows that configuration?

service:
  pipelines:
    traces:
      exporters: [otlp]
    metrics:
      exporters: [otlp]
    logs:
      exporters: [otlp]
```

Replace `gateway-host` with the hostname or IP of your EDOT gateway host. In production, configure TLS to secure communication between the contrib Collector and the gateway.

:::{tip}
Set the `deployment.environment` resource attribute in your contrib Collector so that services appear under the correct environment in the {{kib}} {{product.apm}} Service Map. Without it, all services show as "unset" in the environment selector.

```yaml
processors:
  resource:
    attributes:
      - key: deployment.environment
        action: insert
        value: production
```

Refer to [Attributes and labels](/solutions/observability/apm/opentelemetry/attributes.md) for more details.
:::

> Restart the contrib Collector to apply the changes.

Reviewer: Again, maybe we could use numbered steps for the sub-steps in the stepper?

:::

:::{step} Verify data in {{kib}}

After starting both Collectors, wait a few minutes for data to appear. Then verify in {{kib}}:

> * Navigate to **{{observability}}** → **{{product.apm}}** → **Services** to confirm your services appear.

Reviewer: Same comment for this section: I don't think this wording aligns with our guidance on [referring to apps and pages](https://www.elastic.co/docs/contribute-docs/style-guide/ui-writing#referring-to-apps-and-pages).

* Navigate to **{{observability}}** → **{{product.apm}}** → **Service Map** to confirm environment-based filtering works.
* Navigate to **Discover** and check the `traces-generic.otel-default`, `logs-generic.otel-default`, and `metrics-generic.otel-default` data streams for incoming data.

> If no data appears, refer to [No logs, metrics, or traces visible in {{kib}}](/troubleshoot/ingest/opentelemetry/no-data-in-kibana.md).

Reviewer: What do you think about “… refer to \[…]() for troubleshooting tips” or something similar.

This also doesn’t feel like a nice closure for this guide…  _Next steps_ or _Related pages_ would be a nice addition.

Author: Added "Related pages"

---
navigation_title: Contrib Collector
description: Send data from a contrib OpenTelemetry Collector to a self-managed Elastic Stack by routing it through an EDOT Collector gateway.
applies_to:
  stack: ga 9.2+
products:
  - id: observability
  - id: edot-collector
---

> # Send data from a contrib OpenTelemetry Collector [upstream-collector-self-managed]

Reviewer 2: Why Contrib specifically? OpenTelemetry only has several distros: https://opentelemetry.io/docs/collector/distributions/

I would frame it just `OpenTelemetry Collector (i.e. contrib)` to make it more generic and also cover for the cases where users build their own distros using ocb. We can then say sth like "for the shake of this example we will be using contrib for simplicity".

Author: @ChrsMark you're right, it makes sense to make it clearer. I made some changes to mage it more generic + sprinkled info that this example uses contrib.

> This guide shows how to forward telemetry data from an upstream OpenTelemetry Collector to a self-managed {{stack}} using an EDOT Collector configured as a gateway. The examples use `otelcol-contrib`, but the same approach works with any upstream OTel Collector distribution, including custom builds assembled with the [OpenTelemetry Collector Builder](https://opentelemetry.io/docs/collector/custom-collector/).

Reviewer 3: For clarity, I'd replace `otelcol-contrib` with something like "the popular contrib distribution`, linking to https://github.com/open-telemetry/opentelemetry-collector-releases/tree/main/distributions/otelcol-contrib. The rest of the document can continue using the `otelcol-contrib` shorthand.

## Architecture

Your services send telemetry to the contrib Collector (`otelcol-contrib`), which forwards it over OTLP/gRPC to the EDOT Collector gateway. The gateway applies Elastic-specific processing and writes directly to {{es}}.

```mermaid
flowchart LR
    S["Your services"] -->|OTLP| U["Upstream otelcol-contrib<br/>(one per host)"]
    U -->|"OTLP/gRPC :4317"| G

    subgraph G["EDOT Collector (gateway)"]
        P["elasticapm processor<br/>enriches spans"]
        C["elasticapm connector<br/>generates aggregated metrics"]
    end

    G -->|"elasticsearch exporter<br/>mapping.mode: otel"| E[("ES")]
```

The `elasticsearch` exporter with `mapping.mode: otel` is the recommended path for self-managed deployments. The Managed OTLP endpoint isn't available for self-managed installations. Sending directly to {{apm-server-or-mis}} through OTLP is also possible but not recommended. Use the EDOT gateway path when available.

## Prerequisites

* A running self-managed {{es}} cluster
* The [EDOT Collector](elastic-agent://reference/edot-collector/index.md) installed on the gateway host. It ships as part of the {{agent}} package and runs as {{agent}} in `otel` mode.

> * An upstream OpenTelemetry Collector installed on your agent hosts. This guide uses [`otelcol-contrib`](https://opentelemetry.io/docs/collector/installation/).

Reviewer 3: Technically this doesn't need to be upstream. The user might have a vendor distribution installed, like the [aws otel collector.](https://github.com/aws-observability/aws-otel-collector) Not sure if it's worth making the distinction here.

Reviewer 2: Maybe it could be like "a vanila OpenTelemetry Collector (link to distros) or even a custom built collector". This would cover both the official OpenTelemetry hosted distros but everything else that covers https://github.com/open-telemetry/opentelemetry-collector/tree/main/docs#opentelemetry-collector-distribution

Author: Technically, the contrib/AWS/custom collector still sits _upstream_ in the data flow from the EDOT gateway. We could mention alternatives somewhere near the top of the page (like Christos suggested) to make it clear all of these are an option here.

Some alternatives:

* "an OpenTelemetry Collector" (bit vague)
* "a third-party OpenTelemetry Collector" (smells corporate to me)
* "an existing OpenTelemetry Collector" ("you already have one" use case, but still a bit vague)

I personally like the _"upstream" + explicit mention it can mean vendor/custom/vanilla_ approach the most.

Reviewer 3: "an existing OpenTelemetry Collector" sounds pretty ok to me. If they have an otel collector, they know what it is, and it definitely has an otlp exporter. That's all we care about.

Author: Well that's true. I have to re-read the whole page and think about it some more

::::{stepper}

:::{step} Create an {{es}} API key

The EDOT gateway authenticates to {{es}} using an API key.

1. In {{kib}}, navigate to **{{stack-manage-app}}** → **API keys**.
2. Select **Create API key**.
3. Give the key a name (for example, `edot-gateway`) and assign it the necessary privileges for writing to {{es}} data streams.
4. Copy the encoded key to use as the `ELASTIC_API_KEY` value in the gateway configuration.

:::

:::{step} Configure the EDOT gateway

Create a configuration file for the EDOT Collector acting as a gateway. Save this as `gateway.yml` on the gateway host.

Set the following environment variables on the gateway host before starting the Collector:

```bash
export ELASTIC_ENDPOINT=https://your-elasticsearch:9200
export ELASTIC_API_KEY=your-encoded-api-key
```

Then create `gateway.yml`:

```yaml
receivers:
  otlp:
    protocols:
      grpc:
        endpoint: 0.0.0.0:4317
      http:
        endpoint: 0.0.0.0:4318

connectors:
  elasticapm: {}

processors:
  batch:
    send_batch_size: 1000
    timeout: 1s
    send_batch_max_size: 1500
  batch/metrics:
    send_batch_max_size: 0

>     timeout: 1s

Reviewer 4: Can we get rid of the Batch processor in favor of batching in Elasticsearch exporter?

The Batch processor in current form is buggy. See:
- https://github.com/open-telemetry/opentelemetry-collector/issues/12443
- https://github.com/open-telemetry/opentelemetry-collector/issues/13582
- https://github.com/open-telemetry/opentelemetry-collector/issues/15047

Author: Thanks @andrzej-stencel, that's a good reason to get rid of it. Could you drop the replacement config here? I want to make sure we recommend the right ES exporter batching settings, and I'm not confident enough in the exact parameter names to write it without your confirmation.

Another thing to take into consideration is that with this change we'll diverge from the `elastic-agent/internal/edot/samples/linux/gateway.yml` sample - but I'm guessing that will be updated eventually? In such a case I think it's ok to diverge, even if it's not great to have docs say something different from what you get from the actual EDOT repo...

Author: Hey @andrzej-stencel, could the example ES exporter config look something like this:

```
exporters:
  elasticsearch/otel:
    endpoints:
      - ${env:ELASTIC_ENDPOINT}
    api_key: ${env:ELASTIC_API_KEY}
    mapping:
      mode: otel
    batcher:
      enabled: true
      min_size_items: 1000
      max_size_items: 1500
      flush_timeout: 1s
```
      
I also noticed the original config used a separate batch/metrics with `max_size_items: 0` for metrics pipelines specifically. Is a single batcher config across all signals fine here?

Reviewer 3: It should be fine. For beats, we use presets depending on what the user prefers, but I'm not sure we want to complicate this section too much. I'd also loop in the managed otlp people if we want to be extra careful with this recommendation.

Author: @gizas or @lahsivjar could you please double-check the example `gateway.yml` snippet please? 🙏

Reviewer 5: Alexandra see the docs https://www.elastic.co/docs/reference/edot-collector/components/elasticsearchexporter#performance-and-batching

The bathcer does not seem right? Should be under the sending_queue part. 
I would agree we can have it simpler and refer to this link above maybe

Author: You're right, simpler is always better. I removed the custom batching config entirely and added a cross-link to the performance and batching docs instead. Defaults should be fine for most users

## Prerequisites

* A running self-managed {{es}} cluster
* The [EDOT Collector binary](elastic-agent://reference/edot-collector/index.md) installed on the gateway host
* The [contrib OpenTelemetry Collector](https://opentelemetry.io/docs/collector/installation/) installed on your agent hosts
* Network connectivity from the contrib Collector hosts to the EDOT gateway host on port 4317

::::{stepper}

:::{step} Create an {{es}} API key

The EDOT gateway authenticates to {{es}} using an API key.

1. In {{kib}}, navigate to **{{stack-manage-app}}** → **API keys**.
2. Select **Create API key**.
3. Give the key a name (for example, `edot-gateway`) and assign it the necessary privileges for writing to {{es}} data streams.
4. Copy the encoded key. You will use it as `ELASTIC_API_KEY` in the gateway configuration.

:::

:::{step} Configure the EDOT gateway

Create a configuration file for the EDOT Collector running in gateway mode. Save this as `gateway.yml` on the gateway host.

Set the following environment variables on the gateway host before starting the Collector:

```bash
export ELASTIC_ENDPOINT=https://your-elasticsearch:9200
export ELASTIC_API_KEY=your-encoded-api-key
```

> Then create `gateway.yml`:

Author: I based the gateway.yml on: elastic-agent/internal/edot/samples/linux/gateway.yml

* **`elasticapm` processor** (under `processors`): Enriches spans with attributes required by the {{product.apm}} UI.
* **`elasticapm` connector** (under `connectors`): Generates pre-aggregated {{product.apm}} metrics from trace data. It appears as an exporter in the `traces` pipeline and as a receiver in the `metrics/aggregated-otel-metrics` pipeline.
* **`elasticsearch/otel` exporter**: Writes data directly to {{es}} using native OpenTelemetry data streams (`mapping.mode: otel`).

:::{note}
The `elasticapm` connector and processor are required for full {{product.apm}} functionality (service maps, transaction histograms, service-level indicators). You only need them when exporting directly to {{es}}. If you send to the Managed OTLP endpoint or {{apm-server-or-mis}}, they are not required.

Refer to [{{product.apm}} services missing due to misconfigured elasticapm connector](/troubleshoot/ingest/opentelemetry/edot-collector/misconfigured-elasticapm-connector.md) for more information.
:::

Start the EDOT gateway. The EDOT Collector is the {{agent}} binary run in `otel` mode, so start it with the `otel` subcommand:

```bash
elastic-agent otel --config gateway.yml

> ```

Author: Is this right?

        C["elasticapm connector<br/>generates aggregated metrics"]
    end

    G -->|"elasticsearch exporter<br/>mapping.mode: otel"| E[("Elasticsearch")]
```

The `elasticsearch` exporter with `mapping.mode: otel` is the recommended path for self-managed deployments.

:::{note}
The [Managed OTLP endpoint](opentelemetry://reference/motlp.md) is an alternative ingest path available only on {{ecloud}}, so it doesn't apply to self-managed deployments. Sending directly to {{apm-server-or-mis}} through OTLP is also possible, but the EDOT gateway path in this guide is recommended for full {{product.apm}} functionality.
:::

## Before you begin

> * A running [self-managed](/deploy-manage/deploy/self-managed.md) {{es}} cluster

Reviewer: To transition from the heading to the list, I think it reads better with:

```suggestion
You'll need:

* A running [self-managed](/deploy-manage/deploy/self-managed.md) {{es}} cluster
```
